### PR TITLE
lib: Add contribution guidelines

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

This should be written down so we can refer to it.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by roberth — COMMENTED** [review]
Unfortunately we can't automate ourselves away, but better expectations and structure are great goals.

I just hope people will find it. Maybe link it from relevant other contributing oriented pages? I'm thinking the general coding guidelines, and maybe the PR template?

File: lib/README.md [hunk 1]
```diff
@@ -36,6 +36,24 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
 - [`options.nix`](options.nix): `lib.options` for anything relating to option definitions
 - [`types.nix`](types.nix): `lib.types` for module system types
 
+## Contributing
+
+Follow these guidelines for proposing a change to the interface of `lib`:
+- Motivation: Clearly describe why the change is necessary and its use cases
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply lib/README.md:42]: If your suggestion is to use imperative titles, I believe this is better:

```text
- Do that thing
   
  <More details about why doing that things is good>
```

**Comment by infinisil** [comment]
@roberth I incorporated your feedback, though I did a bit of rewriting for a more consistent style [(diff)](https://github.com/NixOS/nixpkgs/compare/b39453ebafc0ed8e53e2bcaf1fc5d5c26e2c4574..5925f696b4771f1f49f5df15c63d3ba107050639)

File: lib/README.md [hunk 2]
```diff
@@ -36,6 +36,64 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+  make sure to name it appropriately.
+
+  Names should be self-explanatory and consistent with the rest of `lib`.
+  If there's no obvious best name, include the alternatives you considered.
+
+- **Write documentation**
+
+  Update the [reference documentation](#reference-documentation) to reflect the change.
+
+  Be generous with linking to related functionality.
```
Comment (roberth) [inline lib/README.md:70]: nit:

```suggestion
  Be generous with links to related functionality.
```

File: lib/README.md [hunk 3]
```diff
@@ -36,6 +36,64 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+    test messages of custom errors, such as `throw` or `abortMsg`,
+
+    At the time this is only not necessary for sub-libraries tested with [`tests/misc.nix`](./tests/misc.nix).
+
+  See [running tests](#running-tests) for more details on the test suites.
+
+- **Write clean code**
+
+  Name variables well, even if they're internal.
+  Add code comments where the code isn't self-explanatory.
```
Comment (roberth) [inline lib/README.md:89]: We still might want to suggest that self-explanatory code is good, as suggested earlier.

File: lib/README.md [hunk 4]
```diff
@@ -36,6 +36,64 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+
```
Comment (roberth) [inline lib/README.md:90]:
```suggestion
  As a baseline, follow the [Nixpkgs code conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#code-conventions).

```

File: lib/README.md [hunk 5]
```diff
@@ -36,6 +36,64 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+## PR Guidelines
+
+Follow these guidelines for proposing a change to the interface of `lib`.
+
+- **Provide a Motivation**
+
+  Clearly describe why the change is necessary and its use cases.
+
+  Make sure that the change benefits the user more than the added mental effort of looking it up and keeping track of its definition.
+  If the same can be done with the existing interface,
```
Comment (roberth) [inline lib/README.md:48]:
```suggestion
  If the same can reasonably be done with the existing interface,
```

File: lib/README.md [hunk 6]
```diff
@@ -36,6 +36,64 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+  consider just updating the documentation with more examples and links.
+  This is also known as the [Fairbairn Threshold](https://wiki.haskell.org/Fairbairn_threshold).
+
```
Comment (roberth) [inline lib/README.md:51]:
```suggestion
  Through this principle we avoid the human cost of duplicated functionality in an overly large library.

```

**Review by roberth — COMMENTED** [review]
Love the new layout.
Should the items be headers though? That way we get to link to them too.

**Comment by infinisil** [comment]
> Should the items be headers though? That way we get to link to them too.

I did try that out briefly but decided against it because I don't want people to just link to one part of it. This section should be short enough that people can read the entire one without focusing on just one aspect.

**Comment by roberth** [comment]
> I don't want people to just link to one part of it.

Why's that?



> This section should be short

I agree, but seems quite unrelated.

**Comment by infinisil** [comment]
> > I don't want people to just link to one part of it.
> 
> Why's that?

Hmm I didn't think it's needed, but I guess I don't mind very much, I'll change that.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Actually would be good to have links. I mentioned it in the review summary/preface thingy, which is too easy to overlook. Bad UI.

Maybe link it from relevant other contributing oriented pages? I'm thinking the general coding guidelines, and maybe the PR template?

File: lib/README.md [hunk 7]
```diff
@@ -36,6 +36,69 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+Through this principle we avoid the human cost of duplicated functionality in an overly large library.
+
+### Make one PR for each change
+
+Don't have multiple changes in one PR, instead split it up into multiple ones.
+
+This keeps the conversation focused and has a higher chance of getting merged.
+
+### Name the interface appropriately
```
[the remark this answers is not in the document]
Comment (roberth) [reply lib/README.md:60]: If we're going to refer to languages, we'd have to point out the differences as well. That is, link to a document that does that.
Here's a few things:
- Nix lists don't have lazy spines like in Haskell, nor is cons cheap.
- Maybe and Nullable are very different. `mapNullable` is not a lawful functor. It behaves like `>>=` / `flatMap`. `Nullable (Nullable a) == Nullable a`, etc. Less suitable for composition.
- Haskell `Map.union` is flipped compared to `//`.

[the remark this answers is not in the document]
Comment (roberth) [reply lib/README.md:60]: Until we codify anything: maybe make it "here's what we think" and "if you don't know, just open a PR and let's talk about it"?

[the remark this answers is not in the document]
Comment (infinisil) [reply lib/README.md:60]: I added

> If you don't know what name to choose, indicate this as a discussion point in the PR.

to this item. I don't think we could have very concrete guidelines here for now.

File: lib/README.md [hunk 8]
```diff
@@ -36,6 +36,70 @@ The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multi
...
+
+When introducing new names to the interface, such as new function, or new function attributes,
+make sure to name it appropriately.
+
+Names should be self-explanatory and consistent with the rest of `lib`.
+If there's no obvious best name, include the alternatives you considered.
+If you don't know what name to choose, indicate this as a discussion point in the PR.
```
Comment (roberth) [inline lib/README.md:67]: "indicate as discussion point" sounds really heavy.

This sentence also conflicts a bit with the preceding one. Do we really think contributors can't come up with any name whatsoever? It actually seems condescending.

```suggestion
```

or

```suggestion
If you don't know what name to choose, feel free to say so in the PR description.
```

Comment (infinisil) [reply lib/README.md:67]: I don't really mind either way, ping @h7x4

Comment (infinisil) [reply lib/README.md:67]: @h7x4 I was hoping to get some thoughts :sweat_smile:

Comment (infinisil) [reply lib/README.md]: Alright I'll just remove this sentence again. Let's merge before this stalls out for too long. Feel free to open follow-up PRs if there's something to improve.

Comment (h7x4) [reply lib/README.md]: Sorry, the ping went by without me noticing. The thumbs up was intended for the entirety of roberth's comment, so if you removed the sentence I think all is good!

Comment (infinisil) [reply lib/README.md]: Cool thanks!

**Pull request merged**